### ISSUE TITLE
Command to reboot all nodes within the cluster

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#391]
Originally reported of Google Code with ID 335.

```
Feature Request: A Ganeti command to reboot each node within the cluster.

We often perform software upgrades (kernel, Xen, DRBD) that require nodes to be rebooted. It seems logical to have Ganeti determine the best reboot pattern.

It would be nice if the command had to modes: intrusive (requires instance reboot) and non-intrusive. The non-intrusive would perform migrations (and replace-disks?) to ensure the instance is not rebooted.
```
Originally added on 2012-12-18 18:15:14 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Funny you should mention it! Guido already has something (very rought for now), that can compute reboot sets. This will go into 2.7.

After that, we can look into also automating the reboot; should not be a big issue.
```
Originally added on 2012-12-18 19:29:12 +0000 UTC.

Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
To be fair, I had this idea while discussing reboots with Guido.

I was discussing this with Jesus and he raised concerns about using Ganeti to perform reboots when the impetus is Ganeti related. For example, should we use this feature when upgrading xen or Ganeti? How will the final master node reboot be handled?
```
Originally added on 2012-12-19 20:41:03 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
> To be fair, I had this idea while discussing reboots with Guido.

Yeah, he told me afterwards about that :)

> For example, should we use this feature when upgrading xen or Ganeti?

I don't see this as a problem, honestly. There are two parts to this problem:

- computing the reboot sequence; this is safe to do anytime
- actually performing the reboot

The second part is as safe as "gnt-cluster command reboot", which we've used many times and works well. Rebooting the master node is not too special. It will probably not be done from masterd, but as a last step directly from the CLI tool.

So, before: "gnt-cluster command reboot", reboots all at once. After: "hroller reboot --perform" (or whatever the syntax will be), which reboots nodes one by one, or something, and the master at the end.

The tricky bit, I think, will be rather unrelated (to ganeti-or-not-ganeti reboots): how to know for sure a node has come back from reboot completely healthy.
```
Originally added on 2012-12-20 09:43:24 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
We have an additional uses that could benefit from hroller (or a similar tool):

1. OS re-installs
2. Firmware / bios upgrades

We could externalize the drain, update, and reboot steps if hroller can provide the computed reboot sequence (a dry-run mode, basically) in a machine readable format.

Alternatively, hroller could save its state so the reboots may be started, stopped, or controlled.

For example:
hroller reboot --max-reboots 1 --state-dir=/some/path --perform

I'm hoping you have a better idea because I'd rather not duplicate logic externally.
```
Originally added on 2013-01-07 21:50:48 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
hroller is now fully implemented according to its design doc in 2.9
```
Originally added on 2013-07-26 12:44:17 +0000 UTC.

Changed State: Fixed
Added to Milestone: Release2.9

**Comment by Ganeti-Issues-Migrator** [comment]
```
Ganeti 2.9.0 has been released
```
Originally added on 2013-11-05 13:56:36 +0000 UTC.

Changed State: Released
